### ISSUE TITLE
Performance: Loading static data as .js, .json, or JSON.parse

### ISSUE THREAD

**Issue by evelant** [facebook/hermes#1046]
My app has a significant chunk of static data. Currently we load this as an object literal in a ts file. Profiling reveals that loading these static data files takes up a significant chunk of initial load time, at least in profileable environments (android debug builds).

I know that with a "normal" JS runtime it would [likely be faster to JSON.parse the large static object](https://v8.dev/blog/cost-of-javascript-2019) rather than load it as code in a `.js` file. However, since Hermes compiles to bytecode, I'm not sure if that's true in this situation.

So this question is for the experts on Hermes internals (@tmikov ?) -- will it be faster to load a large object/array as a `.js` file or via `JSON.parse` on Hermes?

**Comment by tmikov** [comment]
Unfortunately the answer to this question at the moment is more complicated than I would like it to be.

In Hermes it will always be faster to load it as a .js file instead of parsing it as runtime as JSON. 

*However* Hermes has a long-standing bug, where large literals, depending on their composition, may take extremely long time to compile, or may in fact never complete the compilation successfully. They may also result in extremely inefficient code in debug mode, causing register stack overflow. Note that this is not runtime performance problem, these are design shortcomings in the compiler.

So, **in practice**, for now, it is safer to use `JSON.parse()`, unless you have already tried to compile the large array/literal and seen that it compiles successfully.

We are not proud of this. We finally have a stack of fixes addressing this in a very comprehensive way. This stack will be published in the next major release of Hermes, for which unfortunately we don't have a date yet. We may consider back-porting, depending on the timeline.

So, TLDR. If you can successfully compile it, .js should be significantly faster.

**Comment by evelant** [comment]
Thanks for the comprehensive answer! We haven't run into any compilation failures so I'll keep it as js for now.

**Comment by evelant** [comment]
Somewhat related question regarding perf and Hermes bytecode @tmikov -- does lazy imports of modules likely help or hurt performance on Hermes? I'm guessing this depends on how exactly Hermes handles loading modules and compiling to bytecode.

**Comment by tmikov** [comment]
@evelant another not simple answer :-)

The module system is provided by the packager and implemented in JS, so it isn't Hermes-specific.

Hermes does have its own builtin module system (enabled with the `--commonjs` flag), which is much more efficient and potentially unlocks optimizations like cross-module inlining, but unfortunately it isn't actually used by anyone and thus isn't widely tested. For that reason we decided to stop supporting it and to [disable it in the next point release of Hermes](https://github.com/facebook/hermes/commit/febc8fa57f68309dc2c81d238514ac27cb40dff3).

We will likely re-introduce a zero-cost builtin module system in the next major release, but for now it is better not to think about it.

Back to your question. I am assuming that by *lazy loading of modules* you are referring to the so called *inline requires* transformation. In other words, the `require()` is postponed until the imported module is actually used. The transformation goes from something like this:
```javascript
const mod1 = require("mod1");
...
if (...)
   mod1.foo();
...
mod1.bar(10);
mod2.bar(20);
```
roughly to this:
```javascript
let mod1 = void 0;
...
if (...) 
   (mod1 ? mod1 : (mod1 = require("mod1"))).foo();
...
(mod1 ? mod1 : (mod1 = require("mod1"))).bar(10);
mod1.bar(20);
```

The imported module `mod1` isn't initialized until the first time it is actually used. (I am using CommonJS instead of ESM in this example for simplicity, but in practice packagers lower ESM to CommonJS so it ends up being approximately the same thing.)

Clearly, once a module has been initialized, this adds additional cost to almost every cross-module call.

The win is the delaying of 1) loading, 2) parsing and 3) initialization of the module until it is actually needed, which might even be never. In Hermes, unlike other engines, 1) and 2) are free, so the benefit depends entirely on the cost of 3), initialization.

So, I would say that Hermes benefits less from lazy loading of modules, relatively speaking, since loading of modules is cheaper in Hermes to begin with.
